- Description: zfcp: Error recovery stall when waiting for completion of
               exchange data configuration.
  Symptom:     System hangs.
  Problem:     When the zfcp error recovery retries exchange data configuration
               the corresponding error recovery action structure is not anymore
               enqueued in the running list which leads to a missing wake up
               event for the error recovery thread when the command completes.
  Solution:    Always enqueue corresponding error recovery structure in running
               list of error recovery.
  Problem-ID:  -
  Archs:       s390-31, 390-64

--- linux-2.5/drivers/s390/scsi/zfcp_fsf.c	19 Apr 2004 10:49:37 -0000	1.43.2.2
+++ linux-2.5/drivers/s390/scsi/zfcp_fsf.c	14 May 2004 06:22:23 -0000	1.43.2.3
@@ -29,7 +29,7 @@
  */
 
 /* this drivers version (do not edit !!! generated and updated by cvs) */
-#define ZFCP_FSF_C_REVISION "$Revision: 1.43.2.2 $"
+#define ZFCP_FSF_C_REVISION "$Revision: 1.43.2.3 $"
 
 #include "zfcp_ext.h"
 
@@ -412,7 +412,7 @@
 		fsf_req->status |= ZFCP_STATUS_FSFREQ_ERROR;
 		atomic_set_mask(ZFCP_STATUS_ADAPTER_HOST_CON_INIT,
 				&(adapter->status));
-		debug_text_event(adapter->erp_dbf, 4, "prot_con_init");
+		debug_text_event(adapter->erp_dbf, 3, "prot_con_init");
 		break;
 
 	case FSF_PROT_DUPLICATE_REQUEST_ID:

--- linux-2.5/drivers/s390/scsi/zfcp_erp.c	19 Apr 2004 10:49:37 -0000	1.49.2.2
+++ linux-2.5/drivers/s390/scsi/zfcp_erp.c	14 May 2004 06:22:23 -0000	1.49.2.3
@@ -31,7 +31,7 @@
 #define ZFCP_LOG_AREA			ZFCP_LOG_AREA_ERP
 
 /* this drivers version (do not edit !!! generated and updated by cvs) */
-#define ZFCP_ERP_REVISION "$Revision: 1.49.2.2 $"
+#define ZFCP_ERP_REVISION "$Revision: 1.49.2.3 $"
 
 #include "zfcp_ext.h"
 
@@ -2328,7 +2328,7 @@
 
 	if (qdio_establish(&adapter->qdio_init_data) != 0) {
 		ZFCP_LOG_INFO("error: establishment of QDIO queues failed "
-			      "on adapter %s\n.",
+			      "on adapter %s\n",
 			      zfcp_get_busid_by_adapter(adapter));
 		goto failed_qdio_establish;
 	}
@@ -2540,6 +2540,7 @@
 		atomic_clear_mask(ZFCP_STATUS_ADAPTER_HOST_CON_INIT,
 				  &adapter->status);
 		ZFCP_LOG_DEBUG("Doing exchange config data\n");
+		zfcp_erp_action_to_running(erp_action);
 		zfcp_erp_timeout_init(erp_action);
 		if (zfcp_fsf_exchange_config_data(erp_action)) {
 			retval = ZFCP_ERP_FAILED;
@@ -2566,7 +2567,7 @@
 		 * _must_ be the one belonging to the 'exchange config
 		 * data' request.
 		 */
-		down_interruptible(&adapter->erp_ready_sem);
+		down(&adapter->erp_ready_sem);
 		if (erp_action->status & ZFCP_STATUS_ERP_TIMEDOUT) {
 			ZFCP_LOG_INFO("error: exchange of configuration data "
 				      "for adapter %s timed out\n",

